app/src/full/java/io/homeassistant/companion/android/sensors/HealthConnectSensorManager.kt

```diff
@@ -216,13 +312,179 @@ class HealthConnectSensorManager : SensorManager {
+            response.records.last().distance.inMeters,
+            distance.statelessIcon,
+            attributes = mapOf(
+                "endTime" to response.records.last().endTime,
+                "source" to response.records.last().metadata.dataOrigin.packageName
+            )
+        )
+    }
+
+    private fun updateElevationGainedSensor(context: Context) {
+        val healthConnectClient = getOrCreateHealthConnectClient(context) ?: return
+        val elevationGainedRequest = ReadRecordsRequest(
+            recordType = ElevationGainedRecord::class,
+            timeRangeFilter = TimeRangeFilter.between(
+                Instant.now().minus(30, ChronoUnit.DAYS),
+                Instant.now()
+            ),
+            ascendingOrder = false,
+            pageSize = 1
+        )
+        val response = runBlocking { healthConnectClient.readRecords(elevationGainedRequest) }
+        if (response.records.isEmpty()) {
+            return
+        }
+        onSensorUpdated(
+            context,
+            elevationGained,
+            response.records.last().elevation.inMeters,
+            elevationGained.statelessIcon,
+            attributes = mapOf(
+                "endTime" to response.records.last().endTime,
+                "source" to response.records.last().metadata.dataOrigin.packageName
+            )
+        )
+    }
+
+    private fun updateFloorsClimbedSensor(context: Context) {
+        val healthConnectClient = getOrCreateHealthConnectClient(context) ?: return
+        val floorsClimbedRequest = ReadRecordsRequest(
+            recordType = FloorsClimbedRecord::class,
+            timeRangeFilter = TimeRangeFilter.between(
+                Instant.now().minus(30, ChronoUnit.DAYS),
+                Instant.now()
+            ),
+            ascendingOrder = false,
+            pageSize = 1
+        )
+        val response = runBlocking { healthConnectClient.readRecords(floorsClimbedRequest) }
+        if (response.records.isEmpty()) {
+            return
+        }
+        onSensorUpdated(
+            context,
+            floorsClimbed,
+            response.records.last().floors,
+            floorsClimbed.statelessIcon,
+            attributes = mapOf(
+                "endTime" to response.records.last().endTime,
+                "source" to response.records.last().metadata.dataOrigin.packageName
+            )
+        )
+    }
+
+    private fun updateSleepDurationSensor(context: Context) {
+        val healthConnectClient = getOrCreateHealthConnectClient(context) ?: return
+        val sleepRequest = runBlocking {
+            healthConnectClient.aggregate(
+                AggregateRequest(
+                    metrics = setOf(SleepSessionRecord.SLEEP_DURATION_TOTAL),
+                    timeRangeFilter = TimeRangeFilter.between(
+                        LocalDateTime.of(LocalDate.now(), LocalTime.MIDNIGHT),
+                        LocalDateTime.of(LocalDate.now(), LocalTime.now())
```

Comment:
How does this work if I went to bed at 11 pm?

Reply:
think it depends on teh app recording it, in my own HC data I have gaps in days and some days with double data

Reply:
logic has been changed to pull from the last record instead of aggregate